src/devices/cpu/drcbearm64.cpp

```diff
@@ -0,0 +1,3901 @@
+
+	const uint64_t codeoffs = a.code()->baseAddress() + a.offset();
+	const int64_t reloffs = codeoffs - (int64_t)offs;
+	if (is_valid_immediate_signed(reloffs, 26))
+	{
+		a.bl(offs);
+	}
+	else
+	{
+		get_imm_relative(a, SCRATCH_REG1, uintptr_t(offs));
+		a.blr(SCRATCH_REG1);
+	}
+
+	if (!naked)
+		a.ldp(a64::x29, a64::x30, arm::Mem(a64::sp).post(16));
+}
+
+void drcbe_arm64::check_unordered_condition(a64::Assembler &a, uml::condition_t cond, Label condition_met, bool not_equal) const
+{
+	if (cond != uml::COND_U && cond != uml::COND_NU)
+		return;
+
+	get_unordered(a, SCRATCH_REG1);
+
+	if (cond == uml::COND_U)
+	{
+		if (not_equal)
+			a.cbz(SCRATCH_REG1, condition_met);
+		else
+			a.cbnz(SCRATCH_REG1, condition_met);
+	}
+	else
+	{
+		if (not_equal)
+			a.cbnz(SCRATCH_REG1, condition_met);
+		else
+			a.cbz(SCRATCH_REG1, condition_met);
+	}
+}
+
+void drcbe_arm64::store_unordered(a64::Assembler &a) const
+{
+	a.cset(SCRATCH_REG1, a64::CondCode::kPL);
+	a.cset(SCRATCH_REG2, a64::CondCode::kNE);
+	a.and_(SCRATCH_REG1, SCRATCH_REG1, SCRATCH_REG2);
+	a.cset(SCRATCH_REG2, a64::CondCode::kCS);
+	a.and_(SCRATCH_REG1, SCRATCH_REG1, SCRATCH_REG2);
+	a.cset(SCRATCH_REG2, a64::CondCode::kVS);
+	a.and_(SCRATCH_REG1, SCRATCH_REG1, SCRATCH_REG2);
+	a.bfi(FLAGS_REG, SCRATCH_REG2, 4, 1);
+}
+
+void drcbe_arm64::get_unordered(a64::Assembler &a, const a64::Gp &reg) const
+{
+	get_shifted_bit(a, reg.x(), FLAGS_REG, 1, 4);
+}
+
+void drcbe_arm64::store_carry_reg(a64::Assembler &a, const a64::Gp &reg) const
+{
+	a.bfi(FLAGS_REG, reg.x(), 0, 1);
+}
+
+void drcbe_arm64::store_carry(a64::Assembler &a, bool inverted) const
+{
+	if (inverted)
+		a.cset(SCRATCH_REG1, a64::CondCode::kCC);
+	else
+		a.cset(SCRATCH_REG1, a64::CondCode::kCS);
+
+	store_carry_reg(a, SCRATCH_REG1);
+}
+
+void drcbe_arm64::get_carry(a64::Assembler &a, const a64::Gp &reg, bool inverted) const
+{
+	a.and_(reg.x(), FLAGS_REG, 1);
+
+	if (inverted)
+		a.eor(reg.x(), reg.x(), 1);
+}
+
+void drcbe_arm64::load_carry(a64::Assembler &a, bool inverted) const
+{
+	a.mrs(SCRATCH_REG1, a64::Predicate::SysReg::kNZCV);
+	a.bfi(SCRATCH_REG1, FLAGS_REG, 29, 1);
+
+	if (inverted)
+		a.eor(SCRATCH_REG1, SCRATCH_REG1, 1 << 29);
+
+	a.msr(a64::Predicate::SysReg::kNZCV, SCRATCH_REG1);
+}
+
+void drcbe_arm64::get_shifted_bit(a64::Assembler &a, const a64::Gp &dst, const a64::Gp &src, uint32_t bits, uint32_t shift) const
+{
+	a.lsr(dst.x(), src.x(), shift);
+	a.and_(dst.x(), dst.x(), bits);
+}
+
+void drcbe_arm64::calculate_carry_shift_left(a64::Assembler &a, const a64::Gp &reg, const a64::Gp &shift, int maxBits) const
+{
+	Label skip = a.newLabel();
+	a.cbz(shift, skip);
+
+	const a64::Gp scratch = select_register(SCRATCH_REG1, reg.isGpW() ? 4 : 8);
+	const a64::Gp scratch2 = select_register(SCRATCH_REG2, reg.isGpW() ? 4 : 8);
+
+	// carry = ((PARAM1 << (s	hift - 1)) >> maxBits) & 1
+	a.sub(scratch, shift, 1);
+	a.lsl(scratch2, reg, scratch);
+	a.lsr(scratch2, scratch2, maxBits);
+	store_carry_reg(a, scratch2);
```

Comment:
Can you simplify this to avoid needing two shifts?  Something like
```c++
	a.movz(scratch, maxBits + 1);
	a.sub(scratch, scratch, shift);
	a.lsr(scratch2, reg, scratch);
	store_carry_reg(a, scratch2);
```

You could use a single scratch register.  The `bfi` instruction shouldn’t care if any source bits are set outside the requested field.